// @progress/kendo-react-treeview/dist/npm/utils/FieldsService.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var consts_1 = require("./consts");
var misc_1 = require("./misc");
/**
 * @hidden
 */
var FieldsService = /** @class */ (function () {
    function FieldsService(treeViewProps) {
        this.expandField = treeViewProps.expandField || consts_1.EXPAND_FIELD;
        this.selectField = treeViewProps.selectField || consts_1.SELECT_FIELD;
        this.hasChildrenField = treeViewProps.hasChildrenField || 'hasChildren';
        this.textField = treeViewProps.textField || 'text';
        this.disableField = treeViewProps.disableField || 'disabled';
        this.checkField = treeViewProps.checkField || consts_1.CHECK_FIELD;
        this.checkIndeterminateField = treeViewProps.checkIndeterminateField || consts_1.CHECK_INDETERMINATE_FIELD;
        this.focusIdField = treeViewProps.focusIdField;
    }
    FieldsService.prototype.expanded = function (item) {
        return misc_1.getNestedValue(this.expandField, item);
    };
    FieldsService.prototype.selected = function (item) {
        return misc_1.getNestedValue(this.selectField, item);
    };
    FieldsService.prototype.text = function (item) {
        return misc_1.getNestedValue(this.textField, item);
    };
    FieldsService.prototype.disabled = function (item) {
        return misc_1.getNestedValue(this.disableField, item);
    };
    FieldsService.prototype.hasChildren = function (item) {
        return misc_1.getNestedValue(this.hasChildrenField, item);
    };
    FieldsService.prototype.checked = function (item) {
        return misc_1.getNestedValue(this.checkField, item);
    };
    FieldsService.prototype.checkIndeterminate = function (item) {
        return misc_1.getNestedValue(this.checkIndeterminateField, item);
    };
    FieldsService.prototype.focusId = function (item) {
        return this.focusIdField && misc_1.getNestedValue(this.focusIdField, item);
    };
    return FieldsService;
}());
exports.default = FieldsService;
