// @progress/kendo-react-treeview/dist/npm/utils/itemUtils.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
var itemIdUtils_1 = require("./itemIdUtils");
var SortedPublicItemIds_1 = require("./SortedPublicItemIds");
var misc_1 = require("./misc");
/**
 * @hidden
 */
function isItemExpandedAndWithChildren(item, fieldsSvc) {
    return fieldsSvc.expanded(item) && hasChildren(item);
}
exports.isItemExpandedAndWithChildren = isItemExpandedAndWithChildren;
/**
 * @hidden
 */
function resolveItemId(publicId, idField, items) {
    var resolvedIds = resolveItemsIds([publicId], idField, items);
    return resolvedIds.length ? resolvedIds[0] : undefined;
}
exports.resolveItemId = resolveItemId;
/**
 * @hidden
 */
function resolveItemsIds(publicIds, idField, items) {
    var result = [];
    var sortedIds = new SortedPublicItemIds_1.default();
    sortedIds.init(publicIds);
    resolveItemsIdsHelper(sortedIds, idField, items, itemIdUtils_1.EMPTY_ID, result);
    return result;
}
exports.resolveItemsIds = resolveItemsIds;
/**
 * @hidden
 */
function updateItem(items, itemId, update, cloneField) {
    var _a;
    var itemIndex = Number(itemIdUtils_1.getRootParentId(itemId));
    if (itemIndex >= items.length) {
        return items;
    }
    var result = items;
    var item = items[itemIndex];
    // Clone the item before updating it.
    if (!item[cloneField]) {
        result = items.slice();
        item = result[itemIndex] = Object.assign({}, item, (_a = {}, _a[cloneField] = true, _a));
    }
    // Directly update the item.
    if (itemIdUtils_1.isIdZeroLevel(itemId)) {
        update(item);
    }
    else if (item.items) {
        item.items = updateItem(item.items, itemIdUtils_1.getIdWithoutRootParentId(itemId), update, cloneField);
    }
    return result;
}
exports.updateItem = updateItem;
/**
 * @hidden
 */
function hasChildren(item) {
    return Boolean(item.items && item.items.length);
}
exports.hasChildren = hasChildren;
/**
 * @hidden
 */
function isEnabledAndAllParentsEnabled(itemId, items, fieldsSvc) {
    var ids = itemIdUtils_1.getAllShortIds(itemId);
    var currentItems = items;
    for (var i = 0; i < ids.length; i++) {
        var currentItem = currentItems[Number(ids[i])];
        if (fieldsSvc.disabled(currentItem)) {
            return false;
        }
        else {
            currentItems = currentItem.items;
        }
    }
    return true;
}
exports.isEnabledAndAllParentsEnabled = isEnabledAndAllParentsEnabled;
/**
 * @hidden
 */
function getAllDirectIndirectChildrenIds(item, itemId, idField) {
    return idField ? getChildrenPublicIds(item) : getChildrenHierarchicalIndices(item, itemId);
    function getChildrenHierarchicalIndices(startItem, startItemId) {
        var result = [];
        if (startItem.items) {
            for (var i = 0; i < startItem.items.length; i++) {
                var id = itemIdUtils_1.createId(i, startItemId);
                result.push(id);
                result = result.concat(getChildrenHierarchicalIndices(startItem.items[i], id));
            }
        }
        return result;
    }
    function getChildrenPublicIds(startItem) {
        var result = [];
        if (startItem.items) {
            for (var i = 0; i < startItem.items.length; i++) {
                result.push(misc_1.getNestedValue(idField, startItem.items[i]));
                result = result.concat(getChildrenPublicIds(startItem.items[i]));
            }
        }
        return result;
    }
}
exports.getAllDirectIndirectChildrenIds = getAllDirectIndirectChildrenIds;
/**
 * @hidden
 */
function areAllDirectChildrenChecked(item, itemId, idField, check) {
    if (item.items) {
        return item.items.every(function (child, index) {
            return check.indexOf(idField ? misc_1.getNestedValue(idField, child) : itemIdUtils_1.createId(index, itemId)) > -1;
        });
    }
    else {
        return true;
    }
}
exports.areAllDirectChildrenChecked = areAllDirectChildrenChecked;
/**
 * @hidden
 */
function getAllParents(itemId, items) {
    var result = [];
    var ids = itemIdUtils_1.getAllShortIds(itemId);
    var currentItems = items;
    for (var i = 0; i < ids.length - 1; i++) {
        if (!currentItems) {
            break;
        }
        var currentItem = currentItems[Number(ids[i])];
        result.push(currentItem);
        currentItems = currentItem.items;
    }
    return result;
}
exports.getAllParents = getAllParents;
/**
 * @hidden
 */
function removeItem(itemId, items) {
    var result = items.slice();
    if (itemIdUtils_1.isIdZeroLevel(itemId)) {
        result.splice(Number(itemId), 1);
    }
    else {
        var rootParentIndex = Number(itemIdUtils_1.getRootParentId(itemId));
        var rootParent = result[rootParentIndex] = __assign({}, result[rootParentIndex]);
        rootParent.items = removeItem(itemIdUtils_1.getIdWithoutRootParentId(itemId), rootParent.items);
    }
    return result;
}
exports.removeItem = removeItem;
/**
 * @hidden
 */
function addItem(item, operation, targetItemId, items) {
    var result = items.slice();
    if (itemIdUtils_1.isIdZeroLevel(targetItemId)) {
        if (operation === 'child') {
            var targetItem = result[Number(targetItemId)] = __assign({}, result[Number(targetItemId)]);
            if (targetItem.items) {
                targetItem.items = targetItem.items.slice();
                targetItem.items.push(item);
            }
            else {
                targetItem.items = [item];
            }
        }
        else {
            result.splice(Number(targetItemId) + (operation === 'after' ? 1 : 0), 0, item);
        }
    }
    else {
        var rootParentIndex = Number(itemIdUtils_1.getRootParentId(targetItemId));
        var rootParent = result[rootParentIndex] = __assign({}, result[rootParentIndex]);
        rootParent.items = addItem(item, operation, itemIdUtils_1.getIdWithoutRootParentId(targetItemId), rootParent.items);
    }
    return result;
}
exports.addItem = addItem;
function resolveItemsIdsHelper(publicItemIds, idField, items, parentItemId, resolvedIds) {
    for (var i = 0; i < items.length; i++) {
        var item = items[i];
        var id = itemIdUtils_1.createId(i, parentItemId);
        if (publicItemIds.hasId(misc_1.getNestedValue(idField, item))) {
            resolvedIds.push(id);
        }
        if (hasChildren(item)) {
            resolveItemsIdsHelper(publicItemIds, idField, item.items, id, resolvedIds);
        }
    }
}
