// usetheform/build/es/Select.js
function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

import React from "react";
import { withIndex } from "./hoc/withIndex";
import { useField } from "./hooks/useField";
export var Select = withIndex(function Select(_ref) {
  var onFocus = _ref.onFocus,
      onBlur = _ref.onBlur,
      onChange = _ref.onChange,
      children = _ref.children,
      name = _ref.name,
      index = _ref.index,
      validators = _ref.validators,
      asyncValidator = _ref.asyncValidator,
      onValidation = _ref.onValidation,
      onAsyncValidation = _ref.onAsyncValidation,
      resetSyncErr = _ref.resetSyncErr,
      resetAsyncErr = _ref.resetAsyncErr,
      multiple = _ref.multiple,
      value = _ref.value,
      touched = _ref.touched,
      reducers = _ref.reducers,
      innerRef = _ref.innerRef,
      extraProps = _objectWithoutPropertiesLoose(_ref, ["onFocus", "onBlur", "onChange", "children", "name", "index", "validators", "asyncValidator", "onValidation", "onAsyncValidation", "resetSyncErr", "resetAsyncErr", "multiple", "value", "touched", "reducers", "innerRef"]);

  var props = useField({
    onFocus: onFocus,
    onBlur: onBlur,
    onChange: onChange,
    type: "select",
    name: name,
    index: index,
    value: value,
    validators: validators,
    asyncValidator: asyncValidator,
    onValidation: onValidation,
    onAsyncValidation: onAsyncValidation,
    resetSyncErr: resetSyncErr,
    resetAsyncErr: resetAsyncErr,
    touched: touched,
    multiple: multiple,
    reducers: reducers
  });
  return /*#__PURE__*/React.createElement("select", _extends({}, extraProps, props, {
    ref: innerRef
  }), children);
});
