// usetheform/build/es/hooks/useOwnContext.js
import { createContext, useContext } from "react";
export var ContextObject = /*#__PURE__*/createContext();
export var ContextForm = /*#__PURE__*/createContext();
export var usePublicContextForm = function usePublicContextForm() {
  var _useContext = useContext(ContextForm),
      state = _useContext.state,
      reset = _useContext.reset,
      isValid = _useContext.isValid,
      pristine = _useContext.pristine,
      submitted = _useContext.submitted,
      submitAttempts = _useContext.submitAttempts,
      isSubmitting = _useContext.isSubmitting,
      formStatus = _useContext.formStatus,
      dispatch = _useContext.dispatchNewState,
      onSubmitForm = _useContext.onSubmitForm;

  return {
    state: state,
    reset: reset,
    isValid: isValid,
    pristine: pristine,
    submitted: submitted,
    submitAttempts: submitAttempts,
    isSubmitting: isSubmitting,
    formStatus: formStatus,
    dispatch: dispatch,
    onSubmitForm: onSubmitForm
  };
};
export function useOwnContext() {
  var contextObject = useContext(ContextObject);
  var contextForm = useContext(ContextForm);
  var context = contextObject || contextForm;
  return context;
}
