// react-drip-form/lib/types.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.GroupPropTypes = exports.FieldPropTypes = exports.FormPropTypes = exports.DFContextTypes = undefined;

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var DFContextTypes = exports.DFContextTypes = {
  dripForm: _propTypes2.default.bool.isRequired,
  group: _propTypes2.default.shape({
    name: _propTypes2.default.string,
    multiple: _propTypes2.default.bool
  }),
  register: _propTypes2.default.func.isRequired,
  unregister: _propTypes2.default.func.isRequired,
  updateValue: _propTypes2.default.func.isRequired,
  updateTouched: _propTypes2.default.func.isRequired,
  updateDirty: _propTypes2.default.func.isRequired,
  updateValidations: _propTypes2.default.func.isRequired,
  updateNormalizers: _propTypes2.default.func.isRequired,
  updateLabel: _propTypes2.default.func.isRequired,
  updateMessages: _propTypes2.default.func.isRequired,
  validating: _propTypes2.default.arrayOf(_propTypes2.default.string).isRequired,
  values: _propTypes2.default.object.isRequired,
  errors: _propTypes2.default.object.isRequired,
  dirties: _propTypes2.default.array.isRequired,
  touches: _propTypes2.default.array.isRequired
};
/* eslint-disable import/prefer-default-export */
var FormPropTypes = exports.FormPropTypes = {
  children: _propTypes2.default.node,
  values: _propTypes2.default.object.isRequired,
  errors: _propTypes2.default.object.isRequired,
  meta: _propTypes2.default.shape({
    valid: _propTypes2.default.bool.isRequired,
    invalid: _propTypes2.default.bool.isRequired,
    touched: _propTypes2.default.bool.isRequired,
    untouched: _propTypes2.default.bool.isRequired,
    dirty: _propTypes2.default.bool.isRequired,
    pristine: _propTypes2.default.bool.isRequired,
    validating: _propTypes2.default.bool.isRequired
  }).isRequired,
  fields: _propTypes2.default.shape({
    get: _propTypes2.default.func.isRequired,
    set: _propTypes2.default.func.isRequired,
    remove: _propTypes2.default.func.isRequired,
    push: _propTypes2.default.func.isRequired,
    pop: _propTypes2.default.func.isRequired,
    shift: _propTypes2.default.func.isRequired,
    unshift: _propTypes2.default.func.isRequired,
    swap: _propTypes2.default.func.isRequired,
    move: _propTypes2.default.func.isRequired,
    map: _propTypes2.default.func.isRequired,
    forEach: _propTypes2.default.func.isRequired,
    isValid: _propTypes2.default.func.isRequired,
    isValidating: _propTypes2.default.func.isRequired
  }).isRequired,
  handlers: _propTypes2.default.shape({
    onSubmit: _propTypes2.default.func.isRequired,
    onClear: _propTypes2.default.func.isRequired,
    onReset: _propTypes2.default.func.isRequired
  }).isRequired
};

var FieldPropTypes = exports.FieldPropTypes = {
  input: _propTypes2.default.shape({
    value: _propTypes2.default.any,
    onChange: _propTypes2.default.func.isRequired,
    onFocus: _propTypes2.default.func.isRequired,
    onBlur: _propTypes2.default.func.isRequired
  }).isRequired,
  meta: _propTypes2.default.shape({
    error: _propTypes2.default.string,
    errors: _propTypes2.default.arrayOf(_propTypes2.default.string),
    valid: _propTypes2.default.bool.isRequired,
    invalid: _propTypes2.default.bool.isRequired,
    touched: _propTypes2.default.bool.isRequired,
    untouched: _propTypes2.default.bool.isRequired,
    dirty: _propTypes2.default.bool.isRequired,
    pristine: _propTypes2.default.bool.isRequired,
    validating: _propTypes2.default.bool.isRequired
  }).isRequired
};

var GroupPropTypes = exports.GroupPropTypes = {
  children: _propTypes2.default.node.isRequired,
  meta: _propTypes2.default.shape({
    name: _propTypes2.default.string.isRequired,
    label: _propTypes2.default.string,
    error: _propTypes2.default.string,
    errors: _propTypes2.default.arrayOf(_propTypes2.default.string),
    valid: _propTypes2.default.bool.isRequired,
    invalid: _propTypes2.default.bool.isRequired,
    touched: _propTypes2.default.bool.isRequired,
    untouched: _propTypes2.default.bool.isRequired,
    dirty: _propTypes2.default.bool.isRequired,
    pristine: _propTypes2.default.bool.isRequired,
    validating: _propTypes2.default.bool.isRequired
  }).isRequired
};
